spec/index.bs

```diff
@@ -884,20 +884,90 @@ The [=subresources=] of textures included in the views provided to
 {{GPURenderPassColorAttachment/resolveTarget|GPURenderPassColorAttachment.resolveTarget}}
 are considered to be used as [=internal usage/attachment=] for the [=usage scope=] of this render pass.
 
+The <dfn dfn>logical texture subresource extent</dfn> of a [=texture subresource=] is the size of the
+[=texture subresource=] in texels which are visible by shaders. It is calculated by this procedure:
+
+<div algorithm>
+    <dfn abstract-op>Logical texture subresource extent</dfn>(descriptor, mipLevel)
+
+    **Arguments:**
+        - {{GPUTextureDescriptor}} |descriptor|
+        - {{GPUSize32}} |mipLevel|
+
+    **Returns:** {{GPUExtent3DDict}}
+
+    1. Let |extent| be a new {{GPUExtent3DDict}} object.
+    1. If |descriptor|.{{GPUTextureDescriptor/dimension}} is:
+        <dl class="switch">
+            : {{GPUTextureDimension/"1d"}}
+            ::
+                Set |extent|.{{GPUExtent3DDict/width}} to max(1, |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/width=] &Gt; |mipLevel|).
+
+                Set |extent|.{{GPUExtent3DDict/height}} to 1.
+
+                Set |extent|.{{GPUExtent3DDict/depthOrArrayLayers}} to 1.
+
+            : {{GPUTextureDimension/"2d"}}
+            ::
+                Set |extent|.{{GPUExtent3DDict/width}} to max(1, |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/width=] &Gt; |mipLevel|).
+
+                Set |extent|.{{GPUExtent3DDict/height}} to max(1, |descriptor|.{{GPUTextureDescriptor/size}}.[=Extent3D/height=] &Gt; |mipLevel|).
+
+                Set |extent|.{{GPUExtent3DDict/depthOrArrayLayers}} to 1.
```

Comment:
I think this needs to forward the `depthOrArrayLayers` of the texture to be correct for 2D array textures. Same for the physical extent.

Reply:
This algorithm represents the extent of a single subresource. Each slice of an array texture lies within its own subresource. Therefore, the size of the single subresource does not incorporate the number of array slices in the texture.

Reply:
AFAIU this algorithm would be mainly used for computing that the size of lineardata<->texture copies is in bounds. These copies can copy multiple array layers at the same time, so they'd want to know the depthOrArrayLayers of the texture. Maybe this is just a matter of renaming to "logical/physical texture level extent"?

Reply:
It's used in 2 places:
- [validating GPUImageCopyTexture](https://gpuweb.github.io/gpuweb/#abstract-opdef-validating-gpuimagecopytexture). This is saying that if you're copying in/out of a texture, and the texture is either multisampled or a depth-stencil texture, then the command has to copy exactly one full subresource.
- [validating texture copy range](https://gpuweb.github.io/gpuweb/#validating-texture-copy-range). This is saying that if you're copying in/out of a texture, the bounds specified by the copy command can't exceed the bounds of a single subresource.

I think you're right that the group resolved long ago for these copy commands to be able to copy into multiple levels of an array texture, so I think you're right that, ideally, neither of these places would be restricting the copy to only access a single slice in an array texture.

The concept of subresources is used in [other parts of the spec](https://gpuweb.github.io/gpuweb/#usage-scopes):

> The main usage rule is, for any one [subresource](https://gpuweb.github.io/gpuweb/#subresource), its list of [internal usages](https://gpuweb.github.io/gpuweb/#internal-usage) within one [usage scope](https://gpuweb.github.io/gpuweb/#usage-scopes) must be a [compatible usage list](https://gpuweb.github.io/gpuweb/#compatible-usage-list).

So I don't think we can just redefine "subresource" to mean "a specific miplevel but not a specific array slice" - because its definition is load-bearing.

Therefore, I think you're right about naming; this routine can't claim to be related to the concept of subresources at all. So I think the correct names would be something like
- Logical miplevel-specific texture extent
- Physical miplevel-specific texture extent

Reply:
This is a comprehensive explanation, thank you. I don't understand how the group resolution to allow copying multiple layers of array textures never got encoded in the spec, but this PR now does it.